touchforms/backend/persistence.py

```diff
@@ -24,32 +26,120 @@ def restore(sess_id, factory, override_state=None):
         state.update(override_state)
     return factory(**state)
 
-# TODO integrate with real caching framework (django ideally)
-def cache_set(key, value, timeout):
-    with open(cache_path(key), 'w') as f:
-        f.write(json.dumps(value).encode('utf8'))
 
-# TODO integrate with real caching framework (django ideally)
+def cache_set(key, value):
+    if key is None:
+        raise KeyError
+    if settings.USES_POSTGRES:
+        postgres_insert(key, value)
+    else:
+        with open(cache_get_file_path(key), 'w') as f:
+            f.write(json.dumps(value).encode('utf8'))
+
+
 def cache_get(key):
+    if key is None:
+        raise KeyError
+    if settings.USES_POSTGRES:
+        try:
+            return postgres_lookup(key)
+        except KeyError:
+            return cache_get_file(key)
+    else:
+        return cache_get_file(key)
+
+
+def postgres_select(cursor, key):
+    sel_sql = replace_table("SELECT * FROM %(kwarg)s WHERE sess_id=?")
+    sel_params = [str(key)]
+    cursor.execute(sel_sql, sel_params)
+
+
+def postgres_delete(cursor, key):
+    postgres_select(cursor, key)
+    if cursor.rowcount is not 0:
+        del_sql = replace_table("DELETE FROM %(kwarg)s WHERE sess_id=?")
+        del_params = [str(key)]
+        cursor.execute(del_sql, del_params)
+
+
+def postgres_lookup(key):
+    return postgres_helper(postgres_lookup_command, key)
+
+
+def postgres_lookup_command(cursor, key):
+
+    postgres_select(cursor, key)
+
+    if cursor.rowcount is 0:
+        raise KeyError
+    value = cursor.fetchone()[1]
+
+    jsonobj = json.loads(value.decode('utf8'))
+    return jsonobj
+
+
+def postgres_insert(key, value):
+    postgres_helper(postgres_insert_command, key, value)
+
+
+def postgres_insert_command(cursor, key, value):
+
+    postgres_delete(cursor, key)
+
+    ins_sql = replace_table("INSERT INTO %(kwarg)s (sess_id, sess_json) VALUES (?, ?)")
+    ins_params = [str(key), json.dumps(value).encode('utf8')]
```

Comment:
might make more sense to do an update here rather than deleting and reinserting.